generateBins.py

```diff
@@ -1,86 +1,47 @@
 #!/usr/bin/python3
-# GVF Source: Analytics Vidhya
 # generateBins.py
 #
 # This script generates the bins that need to be run to determine the beta values
 import os
 import sys
-
-from pathlib import Path
-import json
-#import rasterio
-from jenkspy import jenks
-import jenkspy
 import numpy as np
-#import matplotlib.pyplot as plt
-#plt.style.use('seaborn-poster')
-#%matplotlib inline
 
 
 # Import our libraries
 sys.path.append(os.path.join(os.path.dirname(__file__), "include"))
 
-from ascFile import *
-from calibrationLib import *
+from include.ascFile import *
+from include.calibrationLib import *
 
 
 # TODO Still need a good way of supplying these
 PFPR_FILE       = "rwa_pfpr2to10.asc"
 POPULATION_FILE = "rwa_population.asc"
-
+TREATMENT_FILE = "rwa_treatment.asc"
 
 # TODO Determine the bins computationally
 # Following bins are for Rwanda
 POPULATION_BINS = [2125, 5640, 8989, 12108, 15577, 20289, 27629, 49378, 95262, 286928]
 
-# getting data ready for binning
-# data should be 1-dimensional array, python list or iterable
-myArray  = np.loadtxt("GIS\\rwa_population.asc", skiprows = 6)
-print (type(myArray))
-print(myArray.ndim)
-print(myArray.shape)
-print(myArray.size)
-array_1d = myArray.flatten()
-print (array_1d)
-print(array_1d.ndim)
-print(array_1d.shape)
-print(array_1d.size)
-#gvf = 0.0
-#nclasses = 20
-
-def goodness_of_variance_fit(array, classes):
-    # get the break points
-    classes = jenkspy.jenks_breaks(array, classes)
-
-    # do the actual classification
-    classified = np.array([classify(i, classes) for i in array])
-
-    # max value of zones
-    maxz = max(classified)
-
-    # nested list of zone indices
-    zone_indices = [[idx for idx, val in enumerate(classified) if zone + 1 == val] for zone in range(maxz)]
-
-    # sum of squared deviations from array mean
-    sdam = np.sum((array - array.mean()) ** 2)
-
-    # sorted polygon stats
-    array_sort = [np.array([array[index] for index in zone]) for zone in zone_indices]
-
-    # sum of squared deviations of class means
-    sdcm = sum([np.sum((classified - classified.mean()) ** 2) for classified in array_sort])
-
-    # goodness of variance fit
-    gvf = (sdam - sdcm) / sdam
-
-    return gvf
-
-
 def process(configuration, gisPath = ""):
     # Load the configuration
     cfg = load_configuration(configuration)
 
+    # getting data ready for binning
+    # data should be 1-dimensional array, python list or iterable
+    myArray = np.loadtxt(POPULATION_FILE, skiprows=6)
+    array_1d = myArray.flatten()
+
     # TODO Add the stuff for the population bins!
+    # GVF Implementation
+    gvf = 0.0
+    nclasses = 20
+    while gvf < 0.8:
+        gvf = goodness_of_variance_fit(array_1d, nclasses)
+        nclasses += 1
+    print("The value of Goodness of Variance fit is:", gvf)
+    if (gvf < 0.7):
+        print("Warning: GVF too low")
```

Comment:
Okay, from the GVF we should know how to bin the population data. However, I don't seem to see that taking place.

Reply:
I think that should be the classes. As we are taking number of classes as 20 (based on hit and trial; we can increment and decrement the number of classes to get the value for perfect fitting number; ~1) and the GVF is greater than .95, which makes the fit as ~Perfect fit. The value of the break points generated using Jenks is stored in classes, so I think the value in classes must determine the values of bins.

Reply:
@PakhiAgarwal Right, off hand it seems like 20 is likely a high bar for the data sets I've seen so far so I'd start the looking at a low value like 5 or so and then proceed from there. Also, I just noticed this, but there's the possibility for an infinite loop in the code right now.

Reply:
Okay. Based on the data, I'll start from 5 upto 20-25 until I get some number approximate to ~1. This will also help to check for infinite loop.

Reply:
@rjzupkoii These are the values I got:

    `5            0.8187998573940524
     6            0.8400063031786603
     7            0.8725093908883502
     8            0.9105305546827702
     9            0.9206628960735237
     10           0.9282909873982712
     11           0.9410801996635164
     12           0.9453292045975631
     13           0.9460764761434216
     14           0.9560381166408211
     15           0.9565395731869651
     16           0.956969066379069
     17           0.9574810318860014
     18           0.9592570348935402
     19           0.9594522531572597
     20           0.9596282367471943
     21           0.9597394491506118
     22           0.9598343385619622
     23           0.9599260836086719
     24           0.9599659577944439
     25           0.9600349654202794`
    
Looking at these values I think 14 or 15 could be a good value.

Reply:
Some of this should be starting to look familiar from the machine learning courses (e.g., iterative optimization algorithms like gradient descent) you have taken since that's effectively what we are doing here. So there's a couple different ways that we can process the values. 

1. Use best GVF value that we got (i.e., 25); however, we have to be careful about overfitting the data and being mindful of our business rules (i.e., minimize the number of bins so that calibration runs faster). So we can rule that approach out.
2. Use the smallest number of bins with a GVF greater than a given value. Since we are looking for GVF > 0.8 it looks like 5 would have been acceptable. However, based upon the knowledge of the data we are working with, 5 bins might calibrate fast, but depending on where the breaks are they could give us some odd results. I'd prefer a few more bins if possible.
3. Use the smallest number of bins returned at the point where the GVF ceases to change in a a meaningful way. Looking at the code it appears that the hundredths place is a good marker for this and if a `abs(gvf(a) - gvf(b)) <= 0.01` we might have found a result that is "good enough". For this dataset that condition would be satisfied when the number of bins goes from 9 to 10, suggesting that either of those values could work. Incidentally 10 bins were also originally fit for the dataset.

Reply:
That's what I was thinking, to use some ML technique. From 5 to 8 there is a drastic difference in value, then from 8 to 11 we can see the similar difference. But from 14 to 24 it remains almost the same, with little changes. I think we can use some algorithm to find the most optimal value. In this case, I think we can take 14 as most optimal because between 14 to 24, it remains almost the same with little difference. 
I'll look into it more. But for now maybe we can use this.

Reply:
@PakhiAgarwal This entire exercise is an application of ML techniques, so we are good there. :) Don't forget that we really do need to avoid overfitting (i.e., too many bins) with these datasets.